lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -215,15 +215,28 @@ SymbolicValue ConstExprFunctionState::computeConstantValue(SILValue value) {
   // If this is a struct extract from a fragile type, then we can return the
   // element being extracted.
   if (auto *sei = dyn_cast<StructExtractInst>(value)) {
-    auto structValue = sei->getOperand();
-    auto val = getConstantValue(structValue);
-    if (val.isConstant())
-      return val.getAggregateValue()[sei->getFieldNo()];
+    auto aggValue = sei->getOperand();
+    auto val = getConstantValue(aggValue);
+    if (val.isConstant()) {
+      if (val.getKind() == SymbolicValue::Aggregate)
+        return val.getAggregateValue()[sei->getFieldNo()];
+      else {
+        // `val` can be an array. e.g. It can have SIL type $Array<Int32> and be
+        // produce by an inst such as:
+        //   %133 = struct_extract %132 : $TensorShape, #TensorShape.dimensions
+        assert(sei->getFieldNo() == 0);
+        return val;
+      }
+    }
     // Even though the entire struct is not a const value, the field we are
     // extracting could be, so we try evaluating just that field.
-    // TODO: should we remove the code path of getConstantValue(structValue)
-    // above?
-    if (auto *si = dyn_cast<StructInst>(structValue)) {
+    //
+    // Note: This is used in the TF deabstraction context. For example, a tensor
+    // shape attribute value is provided via [someStruct.tensorShape], where
+    // someStruct has non-const fields, and due to the wrapping of the
+    // singleton-array, we are not able to propagate the struct element SSA
+    // value to the use of that attribute in a graph_op inst.
+    if (auto *si = dyn_cast<StructInst>(aggValue)) {
```

Comment:
I'm sorry to be stubborn about this, but this is still not the right direction, because constexpr shouldn't be doing this, this should be a deabstraction problem.  Please remove this hunk of the patch (I understand the testcase may not yet pass).

The right fix for this is to enhance the propagateTensorValues code in DA to remove the struct/structextract stuff.  It gets called after SRoA and before constexpr, and is what is already responsible for doing this for  tensor operands.

Reply:
Update: a const involving `#assert` is probably more like:
```swift

struct S {
  let x: Int
  let y: Int

  init(_ x: Int, _ y: Int) {
    self.x = y
    self.y = y
  }
}

@inline(never)
public func foo() -> Int {
  return 42;
}

public func main() {
  let s = S(3, foo())
  let arr = [s.x]
  #assert(arr[0] == 3)
}
```

But it seems currently const eval cannot handle array subscript yet.

Reply:
Back to our discussion last week, this case cannot be const-evaluated **by design** though, since `s` is not a const value. So i agree extending deabstraction is the right direction.

Reply:
Haha, yeah, I forgot about subscript, we should definitely handle that at some point :-)

Agreed, this is an analogous case.  Your most recent comment makes sense, thanks!

Reply:
Thanks for the follow-up discussion! I'll land this patch as is, and send out a follow-up PR to properly extend deabstraction, and remove the hack in tf const expr.